src/go/k8s/internal/client/factory.go

```diff
@@ -0,0 +1,183 @@
+package client
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-logr/logr"
+	"github.com/redpanda-data/common-go/rpadmin"
+	"github.com/redpanda-data/helm-charts/pkg/redpanda"
+	redpandav1alpha1 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha1"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrEmptyBrokerList             = errors.New("empty broker list")
+	ErrEmptyURLList                = errors.New("empty url list")
+	ErrInvalidKafkaClientObject    = errors.New("cannot initialize Kafka API client from given object")
+	ErrInvalidRedpandaClientObject = errors.New("cannot initialize Redpanda Admin API client from given object")
+	ErrUnsupportedSASLMechanism    = errors.New("unsupported SASL mechanism")
+)
+
+// ClientFactory is responsible for creating both high-level and low-level clients used in our
+// controllers.
+//
+// Calling its `Kafka*` methods will initialize a low-level kgo.Client instance
+// based on the connection parameters contained within the corresponding CRD struct passed in
+// at method invocation.
+//
+// Calling its `RedpandaAdmin*` methods will initialize a low-level rpadmin.AdminAPI instance
+// based on the connection parameters contained within the corresponding CRD struct passed in
+// at method invocation.
+//
+// Additionally, high-level clients that act as functional wrappers to the low-level
+// clients to ease high-level operations (i.e. syncing all primitives for a v1alpha2.User) may
+// be initialized through calling the appropriate `*Client` methods.
+type ClientFactory interface {
+	// WithDialer causes the underlying connections to use a different dialer implementation.
+	WithDialer(dialer redpanda.DialContextFunc) ClientFactory
```

Comment:
Does this need to be exposed from the interface itself? I feel like this would always be a construction time decision, no?